src/operator/numpy/np_matmul_op-inl.h

```diff
@@ -0,0 +1,387 @@
+   * \tparam ndim - ndim of output's argument.
+   * \param N - true ndim of output.
+   */
+  mshadow::Shape<ndim>k_shape;
+  for (int i = shape.ndim() - 1, j = N - 1; i >= 0 || j >= 0 ; --i, --j) {
+    if (i >= 0) {
+      k_shape[j] = shape[i];
+    } else {
+      k_shape[j] = 1;
+    }
+  }
+  return k_shape;
+}
+
+template<int ndim>
+mshadow::Shape<ndim> GetBroadcastKernelShape(mshadow::Shape<ndim> in_shape,
+                                             mshadow::Shape<ndim> broadcast_shape,
+                                             int dimstart, int dimend) {
+  /*!
+   * \brief Broadcast in_shape to broadcast_shape in [dimstart, dimend].
+            Make sure that before use this function:
+            If in_shape[i] != broadcast_shape[i], in_shape[i] == 1.
+   * \param N - ndim of both in_shape and broadcast_shape.
+   * \param dimstart start dimension
+   * \param dimend end dimension
+   */
+  CHECK_GE(dimstart, 0) << "dimstart must be >= 0, while received " << dimstart;
+  CHECK_LT(dimend, ndim) << "dimend must be < " << ndim
+                         << ", while received " << dimend;
+  mshadow::Shape<ndim>out_shape(in_shape);
+  for (int i = dimstart; i < dimend; ++i) {
+    out_shape[i] = std::max(in_shape[i], broadcast_shape[i]);
+  }
+  return out_shape;
+}
+
+template<int req>
+struct SumByShape {
+  /*!
+   * \brief 
+   * \param out - output: insert 'value' to 'arr' according to 'index'.
+   * \param a - input: the first argument.
+   * \param b - input: the second argument.
+   * \param ndim - ndim of a, b and output. Because of broadcast, regard their ndim as equal.  
+   */
+  template<typename DType>
+  MSHADOW_XINLINE static void Map(int i, DType* output, DType* input,
+                                  size_t in_size, size_t out_size){
+    // i is the global position in flattened output
+    size_t pos = static_cast<size_t>(i);
+    DType temp = 0;
+    while (pos < in_size) {
+      temp += input[pos];
+      pos += out_size;
+    }
+    KERNEL_ASSIGN(output[i], req, temp);
+  }
+};
+
+template<typename xpu, typename DType>
+inline void MatmulImpl(const OpContext& ctx,
+                       const TBlob& input_a, const TBlob& input_b, 
+                       const OpReqType& req, const TBlob& output,
+                       Tensor<xpu, 1, char> temp_mem,
+                       const size_t ndim, const size_t batch_size,
+                       const size_t bc_size_a, const size_t bc_size_b,
+                       const mxnet::TShape& a_shape,
+                       const mxnet::TShape& b_shape,
+                       const mxnet::TShape& out_shape,
+                       const bool TA, const bool TB) {
+  using namespace mshadow;
+  using namespace mxnet_op;
+    mshadow::Tensor<xpu, 1, DType*> workspace;
+    mshadow::Tensor<xpu, 3, DType> ans, mlhs, mrhs;
+    mshadow::Stream<xpu> *s = ctx.get_stream<xpu>();
+    if (MatmulNeedBroadcast(a_shape, b_shape)) {
```

Comment:
We can add some documentations above to improve readability. For example, we may describe that we are handling `matmul((1, 3, 4, 2), (3, 1, 2, 4))` case where we need to broadcast lhs/rhs and then call `BatchedGEMM`.